prusti-viper/src/encoder/encoder.rs

```diff
@@ -252,10 +263,144 @@ impl<'v, 'tcx> Encoder<'v, 'tcx> {
                 type_vars: vec![],
             });
         }
+
+        if config::enable_purification_optimization() {
+            domains.push(self.axiomatized_function_domain.borrow().clone());
+            let builtin_encoder =  BuiltinEncoder::new();
+            domains.push(builtin_encoder.encode_builtin_domain(BuiltinDomainKind::Nat));
+            domains.push(builtin_encoder.encode_builtin_domain(BuiltinDomainKind::Primitive));
+        }
+
         domains.sort_by_key(|d| d.get_identifier());
         domains
     }
 
+    pub fn encode_axiomatized_pure_function(&self, f: &vir::Function) {
```

Comment:
This is a huge function that could use some simplifying. Does it actually belong to `Encoder`?

Reply:
Move the function to a separate file and simplified a bit.